Neos.ContentRepository.Core/Classes/Feature/ContentGraphAdapterInterface.php

```diff
@@ -0,0 +1,172 @@
+    ): iterable;
+
+    /**
+     * @throws NodeAggregatesTypeIsAmbiguous
+     */
+    public function findNodeAggregateById(
+        NodeAggregateId $nodeAggregateId
+    ): ?NodeAggregate;
+
+    public function findParentNodeAggregateByChildOriginDimensionSpacePoint(
+        NodeAggregateId $childNodeAggregateId,
+        OriginDimensionSpacePoint $childOriginDimensionSpacePoint
+    ): ?NodeAggregate;
+
+    /**
+     * @return iterable<NodeAggregate>
+     */
+    public function findChildNodeAggregates(
+        NodeAggregateId $parentNodeAggregateId
+    ): iterable;
+
+    /**
+     * @return iterable<NodeAggregate>
+     */
+    public function findTetheredChildNodeAggregates(
+        NodeAggregateId $parentNodeAggregateId
+    ): iterable;
+
+    /**
+     */
+    public function getDimensionSpacePointsOccupiedByChildNodeName(
+        NodeName $nodeName,
+        NodeAggregateId $parentNodeAggregateId,
+        OriginDimensionSpacePoint $parentNodeOriginDimensionSpacePoint,
+        DimensionSpacePointSet $dimensionSpacePointsToCheck
+    ): DimensionSpacePointSet;
+
+    /**
+     * A node aggregate may have multiple child node aggregates with the same name
+     * as long as they do not share dimension space coverage
+     *
+     * @return iterable<NodeAggregate>
+     */
+    public function findChildNodeAggregatesByName(
+        NodeAggregateId $parentNodeAggregateId,
+        NodeName $name
+    ): iterable;
+
+    /*
+     * NODES, basically anything you would ask a subgraph
+     */
+
+    /**
+     * Does the subgraph with the provided identity contain any nodes
+     */
+    public function subgraphContainsNodes(
+        DimensionSpacePoint $dimensionSpacePoint
+    ): bool;
+
+    /**
+     * Finds a specified node within a "subgraph"
+     */
+    public function findNodeInSubgraph(
+        DimensionSpacePoint $coveredDimensionSpacePoint,
+        NodeAggregateId $nodeAggregateId
+    ): ?Node;
+
+    public function findParentNodeInSubgraph(
+        DimensionSpacePoint $coveredDimensionSpacePoint,
+        NodeAggregateId $childNodeAggregateId
+    ): ?Node;
+
+    public function findChildNodesInSubgraph(
+        DimensionSpacePoint $coveredDimensionSpacePoint,
+        NodeAggregateId $parentNodeAggregateId
+    ): Nodes;
+
+    public function findChildNodeByNameInSubgraph(
+        DimensionSpacePoint $coveredDimensionSpacePoint,
+        NodeAggregateId $parentNodeAggregateId,
+        NodeName $nodeNamex
+    ): ?Node;
+
+    public function findPreceedingSiblingNodesInSubgraph(
+        DimensionSpacePoint $coveredDimensionSpacePoint,
+        NodeAggregateId $startingSiblingNodeAggregateId
+    ): Nodes;
+
+    public function findSuceedingSiblingNodesInSubgraph(
+        DimensionSpacePoint $coveredDimensionSpacePoint,
+        NodeAggregateId $startingSiblingNodeAggregateId
+    ): Nodes;
+
+    /*
+     * CONTENT STREAMS
+     */
+
+    public function hasContentStream(): bool;
```

Comment:
as outlined by the ContentGraphAdapterProviderInterface I think it should just throw and not even offer us an adapter if there is no content stream?